### PR TITLE
Vertex reconstructor

### PR DESCRIPTION
**Pull request by christophwelling**
Adds a module that reconstructs the horizontal distance and z-coordinate of the neutrino vertex. It also reconstructs the ray type by adding an empty electric field object for the channels that like detected a signal.

### CODE REVIEW

**Comment by cg-laser** [comment]
@christophwelling I was wondering if you could prepare a small presentation describing this module? It is really difficult to review a reconstruction module just by looking at its code. Maybe you can present it next Thursday in the inicemc call?

**Comment by christophwelling** [comment]
Sounds like a good idea. I will prepare a couple slides

File: NuRadioReco/modules/neutrino2DVertexReconstructor.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        station.set_parameter(stnp.vertex_2D_fit, [self.__rec_x, self.__rec_z])
+        for channel_id in self.__channel_ids:
+            ray_type = self.find_ray_type(station, station.get_channel(channel_id))
+            efield = NuRadioReco.framework.electric_field.ElectricField([channel_id], self.__detector.get_relative_position(station.get_id(), channel_id))
+            efield.set_parameter(efp.ray_path_type, ray_type)
+            station.add_electric_field(efield)
+
+        return
+
+    def get_correlation_array_2d(self, x, z):
```
Comment (anelles) [inline NuRadioReco/modules/neutrino2DVertexReconstructor.py:203]: Document functions please.

File: NuRadioReco/modules/neutrino2DVertexReconstructor.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        channel_pos1 = self.__channel_positions[0]
+        channel_pos2 = self.__channel_positions[1]
+        d_hor1 = np.sqrt((x - channel_pos1[0])**2 + (channel_pos1[1])**2)
+        d_hor2 = np.sqrt((x - channel_pos2[0])**2 + (channel_pos2[1])**2)
+        res = self.get_correlation_for_pos(np.array([d_hor1, d_hor2]), z)
+        return res
+
+    def get_correlation_for_pos(self, d_hor, z):
```
Comment (anelles) [inline NuRadioReco/modules/neutrino2DVertexReconstructor.py:211]: Same here.

File: NuRadioReco/modules/neutrino2DVertexReconstructor.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        delta_t = delta_t.astype(float)
+        corr_index = self.__correlation.shape[0]/2 + np.round(delta_t*self.__sampling_rate)
+        res = np.zeros_like(d_hor[0])
+        mask = (~np.isnan(delta_t)) & (corr_index > 0) & (corr_index < self.__correlation.shape[0]) & (~np.isinf(delta_t))
+        corr_index[~mask] = 0
+        res = np.take(self.__correlation, corr_index.astype(int))
+        res[~mask] = 0
+        return res
+
+    def get_signal_travel_time(self, d_hor, z, ray_type, channel_id):
```
Comment (anelles) [inline NuRadioReco/modules/neutrino2DVertexReconstructor.py:225]: Same.

File: NuRadioReco/modules/neutrino2DVertexReconstructor.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        mask[i_z > self.__lookup_table[channel_type][ray_type].shape[1] - 1] = False
+        i_x[~mask] = 0
+        i_z[~mask] = 0
+        indices = np.array([i_x.flatten(), i_z.flatten()])
+        travel_times = self.__lookup_table[channel_type][ray_type][[i_x, i_z]]
+        travel_times[~mask] = np.nan
+        return travel_times
+
+    def find_ray_type(self, station, ch1):
+        corr_range = 50.*units.ns
```
Comment (anelles) [inline NuRadioReco/modules/neutrino2DVertexReconstructor.py:242]: Same here.

**Review by anelles — CHANGES_REQUESTED** [review]
Some more documentation and then this should be fine.

**Comment by christophwelling** [comment]
I added the documentation you asked for.

**Review by cg-laser — APPROVED** [review]

**Review by anelles — APPROVED** [review]

**Pull request merged**